astropy/coordinates/angles.py

```diff
@@ -363,6 +363,30 @@ def do_format(val):
             result = result[()]
         return result
 
+    def _wrap_at(self, wrap_angle):
+        """
+        Implementation that assumes ``angle`` is already validated
+        and that wrapping is inplace.
+        """
+        # Convert the wrap angle and 360 degrees to the native unit of
+        # this Angle, then do all the math on raw Numpy arrays rather
+        # than Quantity objects for speed.
+        a360 = u.degree.to(self.unit, 360.0)
+        wrap_angle = wrap_angle.to_value(self.unit)
+        wrap_angle_floor = wrap_angle - a360
+        self_angle = self.view(np.ndarray)
+        # Do the wrapping, but only if any angles need to be wrapped
+        #
+        # This invalid catch block can be removed when the minimum numpy
+        # version is >= 1.19 (NUMPY_LT_1_19)
+        with np.errstate(invalid='ignore'):
+            wraps = (self_angle - wrap_angle_floor) // a360
+        if np.any(wraps != 0):
+            self_angle -= wraps*a360
+            # Rounding errors can cause problems.
+            self_angle[self_angle >= wrap_angle] -= a360
+            self_angle[self_angle < wrap_angle_floor] += a360
```

Comment:
Haven't checked closely, but can the order of these matter?

Reply:
I'm quite unhappy with needing these in the first place, and indeed wondered whether there could be some edge case, where one would swing back between a case just above and just below the range. I ended up deciding that below range was the worst...

Reply:
That would indeed be problematic. For the correct value of wrap_angle and wrap_angle_floor it shouldn't ever happen without rounding errors... 🙃

Reply:
Yes, the bane of it all!

Reply:
one potential solution, which may be slow, is to use a numpy trigonometric function that has a well-defined branch cut to coerce everything into the range.

something like 
```
import numpy as np
import astropy.units as u

# want to map into range [-180, 180)
(np.arccos(np.cos(3250 * u.deg)) - 180 * u.deg) << u.deg
```

Reply:
That would be far slower. I did consider writing a quick ufunc for this, but it does seem a bit overkill...  After all, in almost all calculations, it is not as if anything goes wrong if the longitude is too large or small.